File: p2p-ws-sdk-ts/demo/center_join.ts
import fs from "node:fs"
import path from "node:path"
import crypto from "node:crypto"

import { decodeFrame, encodeFrame } from "../src/frame.js"
import { loadProtoRoot } from "../src/proto.js"
import { decodeWrapper, encodeWrapper } from "../src/wrapper.js"
import { loadClientConfig, parseIntMaybeHex } from "../src/config.js"
import { rsaOaepSha256Decrypt } from "../src/handshake.js"

const cfgPath = process.argv[2] ?? path.resolve("..", "p2p-ws-protocol", "examples", "center_client.yaml")
const cfg = loadClientConfig(cfgPath)

const root = await loadProtoRoot()
const Hand = root.lookupType("p2pws.Hand")
const HandAckPlain = root.lookupType("p2pws.HandAckPlain")
const CenterHelloBody = root.lookupType("p2pws.CenterHelloBody")
const CenterHello = root.lookupType("p2pws.CenterHello")
const CenterHelloAck = root.lookupType("p2pws.CenterHelloAck")
const GetNode = root.lookupType("p2pws.GetNode")
const GetNodeAck = root.lookupType("p2pws.GetNodeAck")

const magic = parseIntMaybeHex(cfg.magic, 0x1234)
const version = cfg.version ?? 1
const flagsPlain = cfg.flags_plain ?? 4
const flagsEncrypted = cfg.flags_encrypted ?? 5
const maxFramePayload = cfg.max_frame_payload ?? 4 * 1024 * 1024
const cryptoMode = (cfg.crypto_mode ?? "KEYFILE_XOR_RSA_OAEP").trim()

const keyfileAbs = path.resolve(path.dirname(cfgPath), cfg.keyfile_path)
const fd = fs.openSync(keyfileAbs, "r")
const cachePathRaw = (cfg.presence_cache_path ?? "").trim()
const cachePath = cachePathRaw ? path.resolve(path.dirname(cfgPath), cachePathRaw) : ""

function sha256FileHex(p: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const h = crypto.createHash("sha256")
    const s = fs.createReadStream(p)
    s.on("data", (b) => h.update(b))
    s.on("error", reject)
    s.on("end", () => resolve(h.digest("hex")))
  })
}

const keyHex = await sha256FileHex(keyfileAbs)
const expectedKeyHex = (cfg.key_id_sha256_hex ?? "").trim().toLowerCase()
if (expectedKeyHex && expectedKeyHex !== keyHex) {
  throw new Error(`key_id_sha256_hex mismatch: got=${keyHex} expected=${expectedKeyHex}`)
}
const keyId = Buffer.from(keyHex, "hex")

let privateKeyPem: string
let clientPubDer: Buffer
const privPathRaw = (cfg.rsa_private_key_pem_path ?? "").trim()
if (privPathRaw) {
  const privAbs = path.resolve(path.dirname(cfgPath), privPathRaw)
  privateKeyPem = fs.readFileSync(privAbs, "utf-8")
  const privObj = crypto.createPrivateKey(privateKeyPem)
  const pubObj = crypto.createPublicKey(privObj)
  clientPubDer = pubObj.export({ type: "spki", format: "der" }) as Buffer
} else {
  const kp = crypto.generateKeyPairSync("rsa", { modulusLength: 2048 })
  privateKeyPem = kp.privateKey.export({ type: "pkcs8", format: "pem" }) as string
  clientPubDer = kp.publicKey.export({ type: "spki", format: "der" }) as Buffer
}

const nodeId64Str = String(parseInt(cfg.user_id, 10) || 1)

let offset = -1
let stage = 0
let cachedObserved: { transport: string; addr: string } | null = null
let cachedObservedKey = ""
let lastAck: any | null = null
let renewTimer: NodeJS.Timeout | null = null
const renewSec = typeof cfg.renew_seconds === "number" ? cfg.renew_seconds : 0
let renewLeft = typeof cfg.renew_count === "number" ? cfg.renew_count : renewSec > 0 ? 3 : 0
let queriedOnce = false

if (cachePath) {
  try {
    const txt = fs.readFileSync(cachePath, "utf-8")
    const j = JSON.parse(txt) as any
    const t = String(j?.observed_endpoint?.transport ?? "")
    const a = String(j?.observed_endpoint?.addr ?? "")
    if (t && a) {
      cachedObserved = { transport: t, addr: a }
      cachedObservedKey = t + "|" + a
    }
  } catch {}
}

function xorWithFile(data: Uint8Array, off: number): Uint8Array {
  const slice = Buffer.allocUnsafe(data.length)
  let pos = 0
  while (pos < slice.length) {
    const n = fs.readSync(fd, slice, pos, slice.length - pos, off + pos)
    if (n <= 0) throw new Error("read keyfile failed")
    pos += n
  }
  const out = new Uint8Array(data.length)
  for (let i = 0; i < data.length; i++) out[i] = data[i] ^ slice[i]
  return out
}

function applyObservedIpToEndpoints(observedAddr: string, endpoints: any[]): any[] {
  const ip = String(observedAddr).split(":")[0]
  if (!ip) return endpoints
  const out: any[] = []
  for (const e of endpoints) {
    const t = String(e?.transport ?? "")
    const addr = String(e?.addr ?? "")
    const parts = addr.split(":")
    if (parts.length >= 2) {
      out.push({ transport: t, addr: ip + ":" + parts.slice(1).join(":") })
    } else {
      out.push({ transport: t, addr })
    }
  }
  return out
}

const ws = new WebSocket(cfg.ws_url)
ws.binaryType = "arraybuffer"

ws.onopen = () => {
  console.log("open")
  const initialReported = Array.isArray(cfg.reported_endpoints) ? cfg.reported_endpoints.slice() : []
  const initialWithCache = cachedObserved ? applyObservedIpToEndpoints(cachedObserved.addr, initialReported) : initialReported
  const handData = Hand.encode(
    Hand.create({
      client_pubkey: clientPubDer,
      key_ids: [keyId],
      max_frame_payload: maxFramePayload,
      client_id: cfg.user_id,
    }),
  ).finish()
  const wrapper = encodeWrapper(root, { seq: 1, command: -10001, data: handData })
  ws.send(encodeFrame({ length: wrapper.length, magic, version, flags: flagsPlain }, wrapper))
  stage = 1
  ;(ws as any).__reported = initialWithCache
}

ws.onmessage = (ev) => {
  const wsBytes = new Uint8Array(ev.data as ArrayBuffer)
  const f = decodeFrame(wsBytes)
  const cipher = f.cipherPayload
  const plainPayload = offset >= 0 ? xorWithFile(cipher, offset) : cipher
  const w = decodeWrapper(root, plainPayload)
  console.log("recv.cmd=" + w.command)

  if (w.command === -10002) {
    const enc = w.data ?? new Uint8Array()
    const handAckBytes = rsaOaepSha256Decrypt(privateKeyPem, enc)
    const ack = HandAckPlain.decode(handAckBytes) as any
    offset = Number(ack.offset)
    stage = 2

    const caps = { max_frame_payload: maxFramePayload, magic, version, flags_plain: flagsPlain, flags_encrypted: flagsEncrypted }
    const reported = ((ws as any).__reported ?? []) as any[]
    const bodyMsg = CenterHelloBody.create({
      node_id64: nodeId64Str,
      pubkey_spki_der: clientPubDer,
      reported_endpoints: reported,
      caps,
      timestamp_ms: String(Date.now()),
      crypto_mode: cryptoMode,
    })
    const bodyBytes = CenterHelloBody.encode(bodyMsg).finish()
    const sig = crypto.sign("RSA-SHA256", Buffer.from(bodyBytes), crypto.createPrivateKey(privateKeyPem))
    const hello = CenterHello.encode(CenterHello.create({ body: bodyMsg, signature: sig })).finish()

    const helloWrap = encodeWrapper(root, { seq: 2, command: -11001, data: hello })
    const helloCipher = xorWithFile(helloWrap, offset)
    ws.send(encodeFrame({ length: helloCipher.length, magic, version, flags: flagsEncrypted }, helloCipher))
    stage = 3
    return
  }

  if (w.command === -11002) {
    const ack = CenterHelloAck.decode(w.data ?? new Uint8Array()) as any
    lastAck = ack
    const oldKey = cachedObservedKey
    const nodeKey = Buffer.from(ack.node_key as Uint8Array).toString("hex")
    console.log("center_ack.node_key=" + nodeKey)
    if (ack.observed_endpoint) {
      const oe = ack.observed_endpoint as any
      console.log("center_ack.observed_endpoint=" + String(oe.transport) + ":" + String(oe.addr))
      const oeKey = String(oe.transport) + "|" + String(oe.addr)
      cachedObservedKey = oeKey
      cachedObserved = { transport: String(oe.transport), addr: String(oe.addr) }
      if (cachePath && oe.transport && oe.addr) {
        try {
          fs.mkdirSync(path.dirname(cachePath), { recursive: true })
          fs.writeFileSync(cachePath, JSON.stringify({ observed_endpoint: { transport: String(oe.transport), addr: String(oe.addr) } }, null, 2), "utf-8")
        } catch {}
      }
    }
    stage = 4
    if (ack.observed_endpoint) {
      const oe = ack.observed_endpoint as any
      const oeKey = String(oe.transport) + "|" + String(oe.addr)
      if (!oldKey || oldKey !== oeKey) {
        const caps = { max_frame_payload: maxFramePayload, magic, version, flags_plain: flagsPlain, flags_encrypted: flagsEncrypted }
        const base = ((ws as any).__reported ?? []) as any[]
        const applied = applyObservedIpToEndpoints(String(oe.addr), base)
        const bodyMsg2 = CenterHelloBody.create({
          node_id64: nodeId64Str,
          pubkey_spki_der: clientPubDer,
          reported_endpoints: applied,
          caps,
          timestamp_ms: String(Date.now()),
          crypto_mode: cryptoMode,
        })
        const b2 = CenterHelloBody.encode(bodyMsg2).finish()
        const sig2 = crypto.sign("RSA-SHA256", Buffer.from(b2), crypto.createPrivateKey(privateKeyPem))
        const hello2 = CenterHello.encode(CenterHello.create({ body: bodyMsg2, signature: sig2 })).finish()
        const wrap2 = encodeWrapper(root, { seq: 21, command: -11001, data: hello2 })
        const cipherX = xorWithFile(wrap2, offset)
        ws.send(encodeFrame({ length: cipherX.length, magic, version, flags: flagsEncrypted }, cipherX))
      }
    }

    if (!queriedOnce) {
      queriedOnce = true
      const req = GetNode.encode(GetNode.create({ node_key: ack.node_key })).finish()
      const wrap = encodeWrapper(root, { seq: 3, command: -11010, data: req })
      const cipher2 = xorWithFile(wrap, offset)
      ws.send(encodeFrame({ length: cipher2.length, magic, version, flags: flagsEncrypted }, cipher2))
    }
    return
  }

  if (w.command === -11011) {
    const r = GetNodeAck.decode(w.data ?? new Uint8Array()) as any
    console.log("get_node.found=" + String(r.found))
    if (renewSec <= 0) {
      fs.closeSync(fd)
      ws.close()
      stage = 5
    } else {
      stage = 6
    }
  }
}

ws.onerror = (e) => {
  console.error(e)
  process.exit(1)
}

if (renewSec <= 0) {
  setTimeout(() => {
    if (stage < 5) {
      console.error("timeout.stage=" + stage)
      try {
        fs.closeSync(fd)
      } catch {}
      ws.close()
      process.exit(2)
    }
  }, 8000)
}

function sendGetNode() {
  if (!lastAck?.node_key || offset < 0) return
  const req = GetNode.encode(GetNode.create({ node_key: lastAck.node_key })).finish()
  const wrap = encodeWrapper(root, { seq: 300, command: -11010, data: req })
  const cipher = xorWithFile(wrap, offset)
  ws.send(encodeFrame({ length: cipher.length, magic, version, flags: flagsEncrypted }, cipher))
}

function finishAfterExpiry() {
  const ttl = Number(lastAck?.ttl_seconds ?? 0)
  if (ttl <= 0) return
  setTimeout(() => {
    sendGetNode()
    setTimeout(() => {
      try {
        fs.closeSync(fd)
      } catch {}
      ws.close()
      process.exit(0)
    }, 1500)
  }, (ttl + 1) * 1000)
}

function sendRenew() {
  if (stage < 2 || offset < 0) return
  if (!lastAck?.node_key) return
  const caps = { max_frame_payload: maxFramePayload, magic, version, flags_plain: flagsPlain, flags_encrypted: flagsEncrypted }
  const reported = ((ws as any).__reported ?? []) as any[]
  const bodyMsg = CenterHelloBody.create({
    node_id64: nodeId64Str,
    pubkey_spki_der: clientPubDer,
    reported_endpoints: reported,
    caps,
    timestamp_ms: String(Date.now()),
    crypto_mode: cryptoMode,
  })
  const bodyBytes = CenterHelloBody.encode(bodyMsg).finish()
  const sig = crypto.sign("RSA-SHA256", Buffer.from(bodyBytes), crypto.createPrivateKey(privateKeyPem))
  const hello = CenterHello.encode(CenterHello.create({ body: bodyMsg, signature: sig })).finish()
  const wrap = encodeWrapper(root, { seq: 100 + renewLeft, command: -11001, data: hello })
  const cipher = xorWithFile(wrap, offset)
  ws.send(encodeFrame({ length: cipher.length, magic, version, flags: flagsEncrypted }, cipher))
  if (renewLeft > 0) {
    renewLeft -= 1
    if (renewLeft === 0 && renewTimer) {
      clearInterval(renewTimer)
      renewTimer = null
      finishAfterExpiry()
    }
  }
}

if (renewSec > 0) {
  renewTimer = setInterval(sendRenew, renewSec * 1000)
}

File: p2p-ws-sdk-python/demo/center_join.py
from __future__ import annotations

import asyncio
import hashlib
from pathlib import Path
import sys

root = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(root / "src"))

from p2p_ws_sdk.frame import decode_frame, encode_frame, WireHeader
from p2p_ws_sdk.xor import xor_no_wrap


async def main() -> int:
    try:
        import websockets  # type: ignore
    except Exception:
        raise SystemExit("websockets is required: python -m pip install websockets")

    try:
        import yaml  # type: ignore
    except Exception:
        raise SystemExit("PyYAML is required: python -m pip install PyYAML")

    try:
        from cryptography.hazmat.primitives.asymmetric import padding  # type: ignore
        from cryptography.hazmat.primitives import hashes, serialization  # type: ignore
    except Exception:
        raise SystemExit("cryptography is required: python -m pip install cryptography")

    try:
        from p2p_ws_sdk.gen import p2p_wrapper_pb2, p2p_control_pb2  # type: ignore
    except Exception:
        raise SystemExit("pb2 is required: python p2p-ws-sdk-python/scripts/gen_proto.py")

    cfg_path = Path(sys.argv[1]) if len(sys.argv) >= 2 else (root.parent / "p2p-ws-protocol" / "examples" / "center_client.yaml")
    cfg = yaml.safe_load(cfg_path.read_text(encoding="utf-8"))

    ws_url = str(cfg.get("ws_url"))
    user_id_raw = cfg.get("user_id")
    node_id64 = int(str(user_id_raw), 10) if user_id_raw is not None else 0
    keyfile_path = Path(str(cfg.get("keyfile_path")))
    if not keyfile_path.is_absolute():
        keyfile_path = (cfg_path.parent / keyfile_path).resolve()

    kid_v = cfg.get("key_id_sha256_hex", "")
    expected_key_id_hex = ("" if kid_v is None else str(kid_v)).strip().lower()
    pem_v = cfg.get("rsa_private_key_pem_path", "")
    rsa_private_key_pem_path = ("" if pem_v is None else str(pem_v)).strip()

    magic = int(str(cfg.get("magic", "0x1234")), 0) if isinstance(cfg.get("magic", None), str) else int(cfg.get("magic", 0x1234))
    version = int(cfg.get("version", 1))
    flags_plain = int(cfg.get("flags_plain", 4))
    flags_encrypted = int(cfg.get("flags_encrypted", 5))
    max_frame_payload = int(cfg.get("max_frame_payload", 4 * 1024 * 1024))
    crypto_mode = str(cfg.get("crypto_mode", "KEYFILE_XOR_RSA_OAEP")).strip()
    reported_endpoints_cfg = cfg.get("reported_endpoints", [])
    cache_v = cfg.get("presence_cache_path", "")
    presence_cache_path = ("" if cache_v is None else str(cache_v)).strip()
    renew_seconds = int(cfg.get("renew_seconds", 0) or 0)
    renew_count = int(cfg.get("renew_count", 0) or 0)
    if renew_seconds > 0 and renew_count <= 0:
        renew_count = 3

    h = hashlib.sha256()
    with keyfile_path.open("rb") as f:
        while True:
            b = f.read(1024 * 1024)
            if not b:
                break
            h.update(b)
    key_id = h.digest()
    if expected_key_id_hex and key_id.hex() != expected_key_id_hex:
        raise SystemExit(f"key_id_sha256_hex mismatch: got={key_id.hex()} expected={expected_key_id_hex}")

    if not rsa_private_key_pem_path:
        raise SystemExit("rsa_private_key_pem_path required for center join demo")
    pem_path = Path(rsa_private_key_pem_path)
    if not pem_path.is_absolute():
        pem_path = (cfg_path.parent / pem_path).resolve()
    priv = serialization.load_pem_private_key(pem_path.read_bytes(), password=None)
    pub = priv.public_key()
    pub_der = pub.public_bytes(encoding=serialization.Encoding.DER, format=serialization.PublicFormat.SubjectPublicKeyInfo)

    node_key = hashlib.sha256(pub_der).digest()

    async with websockets.connect(ws_url) as ws:
        keyf = keyfile_path.open("rb")
        cached_observed = None
        cached_observed_key = ""
        if presence_cache_path:
            p = Path(presence_cache_path)
            if not p.is_absolute():
                p = (cfg_path.parent / p).resolve()
            try:
                j = __import__("json").loads(p.read_text(encoding="utf-8"))
                oe = j.get("observed_endpoint") if isinstance(j, dict) else None
                if isinstance(oe, dict) and oe.get("transport") and oe.get("addr"):
                    cached_observed = {"transport": str(oe["transport"]), "addr": str(oe["addr"])}
                    cached_observed_key = cached_observed["transport"] + "|" + cached_observed["addr"]
            except Exception:
                cached_observed = None

        hand = p2p_control_pb2.Hand(client_pubkey=pub_der, key_ids=[key_id], max_frame_payload=max_frame_payload, client_id=str(node_id64))
        w = p2p_wrapper_pb2.P2PWrapper(seq=1, command=-10001, data=hand.SerializeToString())
        plain = w.SerializeToString()
        await ws.send(encode_frame(WireHeader(len(plain), magic, version, flags_plain), plain))

        offset: int | None = None

        def xor_with_file(data: bytes, off: int) -> bytes:
            keyf.seek(off)
            slice_ = keyf.read(len(data))
            if len(slice_) != len(data):
                raise RuntimeError("read keyfile failed")
            return xor_no_wrap(data, slice_, 0)

        def apply_observed_ip(observed_addr: str, endpoints: list) -> list:
            ip = str(observed_addr).split(":")[0]
            if not ip:
                return endpoints
            out = []
            for e in endpoints:
                if not isinstance(e, dict):
                    continue
                t = str(e.get("transport", ""))
                addr = str(e.get("addr", ""))
                parts = addr.split(":")
                if len(parts) >= 2:
                    out.append({"transport": t, "addr": ip + ":" + ":".join(parts[1:])})
                else:
                    out.append({"transport": t, "addr": addr})
            return out

        ttl_from_ack = 0
        expect_expired = False
        queried_once = False

        async def send_hello(seq: int, reported: list) -> None:
            nonlocal offset
            if offset is None:
                return
            caps = p2p_control_pb2.NodeCaps(
                max_frame_payload=max_frame_payload,
                magic=magic,
                version=version,
                flags_plain=flags_plain,
                flags_encrypted=flags_encrypted,
            )
            body = p2p_control_pb2.CenterHelloBody(
                node_id64=node_id64,
                pubkey_spki_der=pub_der,
                reported_endpoints=reported,
                caps=caps,
                timestamp_ms=int(__import__("time").time() * 1000),
                crypto_mode=crypto_mode,
            )
            sig = priv.sign(body.SerializeToString(), padding.PKCS1v15(), hashes.SHA256())
            hello = p2p_control_pb2.CenterHello(body=body, signature=sig)
            wrap = p2p_wrapper_pb2.P2PWrapper(seq=seq, command=-11001, data=hello.SerializeToString())
            wp = wrap.SerializeToString()
            c = xor_with_file(wp, offset)
            await ws.send(encode_frame(WireHeader(len(c), magic, version, flags_encrypted), c))

        async def send_get_node(seq: int) -> None:
            nonlocal offset
            if offset is None:
                return
            req = p2p_control_pb2.GetNode(node_id64=node_id64, node_key=node_key)
            wrap = p2p_wrapper_pb2.P2PWrapper(seq=seq, command=-11010, data=req.SerializeToString())
            wp = wrap.SerializeToString()
            c = xor_with_file(wp, offset)
            await ws.send(encode_frame(WireHeader(len(c), magic, version, flags_encrypted), c))

        while True:
            msg = await ws.recv()
            if not isinstance(msg, (bytes, bytearray)):
                continue
            f = decode_frame(bytes(msg))
            payload = f.cipher_payload
            plain_payload = payload if offset is None else xor_with_file(payload, offset)
            back = p2p_wrapper_pb2.P2PWrapper()
            back.ParseFromString(plain_payload)

            if back.command == -10002:
                decrypted = priv.decrypt(
                    back.data,
                    padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None),
                )
                ack = p2p_control_pb2.HandAckPlain()
                ack.ParseFromString(decrypted)
                offset = int(ack.offset)

                caps = p2p_control_pb2.NodeCaps(
                    max_frame_payload=max_frame_payload,
                    magic=magic,
                    version=version,
                    flags_plain=flags_plain,
                    flags_encrypted=flags_encrypted,
                )
                reported = reported_endpoints_cfg if isinstance(reported_endpoints_cfg, list) else []
                if cached_observed:
                    reported = apply_observed_ip(cached_observed.get("addr", ""), list(reported))
                body = p2p_control_pb2.CenterHelloBody(
                    node_id64=node_id64,
                    pubkey_spki_der=pub_der,
                    reported_endpoints=reported,
                    caps=caps,
                    timestamp_ms=int(__import__("time").time() * 1000),
                    crypto_mode=crypto_mode,
                )
                sig = priv.sign(body.SerializeToString(), padding.PKCS1v15(), hashes.SHA256())
                hello = p2p_control_pb2.CenterHello(body=body, signature=sig)

                wrap = p2p_wrapper_pb2.P2PWrapper(seq=2, command=-11001, data=hello.SerializeToString())
                wrap_plain = wrap.SerializeToString()
                cipher = xor_with_file(wrap_plain, offset)
                await ws.send(encode_frame(WireHeader(len(cipher), magic, version, flags_encrypted), cipher))
                continue

            if back.command == -11002:
                ack = p2p_control_pb2.CenterHelloAck()
                ack.ParseFromString(back.data)
                if ack.node_key != node_key:
                    raise SystemExit("node_key mismatch in ack")
                ttl_from_ack = int(ack.ttl_seconds)
                if ack.observed_endpoint and ack.observed_endpoint.addr:
                    print("center_ack.observed_endpoint=" + str(ack.observed_endpoint.transport) + ":" + str(ack.observed_endpoint.addr))
                    cached_observed_key = str(ack.observed_endpoint.transport) + "|" + str(ack.observed_endpoint.addr)
                    cached_observed = {"transport": str(ack.observed_endpoint.transport), "addr": str(ack.observed_endpoint.addr)}
                    if presence_cache_path:
                        p = Path(presence_cache_path)
                        if not p.is_absolute():
                            p = (cfg_path.parent / p).resolve()
                        try:
                            p.parent.mkdir(parents=True, exist_ok=True)
                            p.write_text(
                                __import__("json").dumps(
                                    {"observed_endpoint": {"transport": str(ack.observed_endpoint.transport), "addr": str(ack.observed_endpoint.addr)}},
                                    indent=2,
                                ),
                                encoding="utf-8",
                            )
                        except Exception:
                            pass

                if offset is not None and ack.observed_endpoint and ack.observed_endpoint.addr:
                    oe_key = str(ack.observed_endpoint.transport) + "|" + str(ack.observed_endpoint.addr)
                    if not (cached_observed_key and cached_observed_key == oe_key):
                        caps = p2p_control_pb2.NodeCaps(
                            max_frame_payload=max_frame_payload,
                            magic=magic,
                            version=version,
                            flags_plain=flags_plain,
                            flags_encrypted=flags_encrypted,
                        )
                        base_reported = reported_endpoints_cfg if isinstance(reported_endpoints_cfg, list) else []
                        applied = apply_observed_ip(str(ack.observed_endpoint.addr), list(base_reported))
                        body2 = p2p_control_pb2.CenterHelloBody(
                            node_id64=node_id64,
                            pubkey_spki_der=pub_der,
                            reported_endpoints=applied,
                            caps=caps,
                            timestamp_ms=int(__import__("time").time() * 1000),
                            crypto_mode=crypto_mode,
                        )
                        sig2 = priv.sign(body2.SerializeToString(), padding.PKCS1v15(), hashes.SHA256())
                        hello2 = p2p_control_pb2.CenterHello(body=body2, signature=sig2)
                        wrap2 = p2p_wrapper_pb2.P2PWrapper(seq=21, command=-11001, data=hello2.SerializeToString())
                        wrap_plain2 = wrap2.SerializeToString()
                        cipher2 = xor_with_file(wrap_plain2, offset)
                        await ws.send(encode_frame(WireHeader(len(cipher2), magic, version, flags_encrypted), cipher2))

                if not queried_once:
                    queried_once = True
                    req = p2p_control_pb2.GetNode(node_id64=node_id64, node_key=ack.node_key)
                    wrap = p2p_wrapper_pb2.P2PWrapper(seq=3, command=-11010, data=req.SerializeToString())
                    wrap_plain = wrap.SerializeToString()
                    cipher = xor_with_file(wrap_plain, offset if offset is not None else 0)
                    await ws.send(encode_frame(WireHeader(len(cipher), magic, version, flags_encrypted), cipher))
                continue

            if back.command == -11011:
                r = p2p_control_pb2.GetNodeAck()
                r.ParseFromString(back.data)
                print("get_node.found=" + str(bool(r.found)).lower())
                if renew_seconds > 0 and not expect_expired:
                    expect_expired = True

                    async def run_and_expire() -> None:
                        reported0 = reported_endpoints_cfg if isinstance(reported_endpoints_cfg, list) else []
                        if cached_observed:
                            reported0 = apply_observed_ip(cached_observed.get("addr", ""), list(reported0))
                        for i in range(renew_count):
                            await asyncio.sleep(renew_seconds)
                            await send_hello(100 + i, reported0)
                        wait_s = (ttl_from_ack + 1) if ttl_from_ack > 0 else 2
                        await asyncio.sleep(wait_s)
                        await send_get_node(300)

                    asyncio.create_task(run_and_expire())
                    continue

                keyf.close()
                return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
